### PR TITLE
BFD-3672: Populate tags for claim / EOB on ingestion from RDA

### PR DESCRIPTION
**Pull request by dondevun**
**JIRA Ticket:**
BFD-3672

### What Does This PR Do?

This is the first stage in the creation of the SAMHSA filter 2.0. This creates the tags for the RDA claims, and creates a new class, SamhsaUtil. This class will allow for the evaluation of the claims for SAMHSA codes as they get saved to the database, and persists the associated SAMHSA tags, if necessary.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

### What Security Implications Does This PR Have?

Please indicate if this PR does any of the following:  

* Adds any new software dependencies
* Modifies any security controls
* Adds new transmission or storage of data
* Any other changes that could possibly affect security? 

* [x] I have considered the above security implications as it relates to this PR. (If one or more of the above apply, it cannot be merged without the ISSO or team security engineer's (`@sb-benohe`) approval.) 

### Validation

Tested by running the Pipeline locally. Additionally, Integration tests and Unit tests have been created.

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+    return entries.isEmpty() ? Optional.empty() : Optional.of(entries);
+  }
+
+  /**
+   * Check if a given code is a SAMHSA code.
+   *
+   * @param code the code to check.
+   * @return If the code is SAMHSA, returns the SAMHSA entry. Otherwise, an empty optional.
+   */
+  public Optional<SamhsaEntry> isSamhsaCode(Optional<String> code) {
```
Comment (aschey-forpeople) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java:361]: Maybe this should be called `getSamhsaCode`? `isSamhsaCode` implies that it returns a boolean.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java [hunk 2]
```diff
@@ -408,6 +410,7 @@ private void mergeBatch(long maxSeq, Collection<RdaChange<TClaim>> changes) {
                 var metaData = createMetaData(change);
                 entityManager.merge(metaData);
                 entityManager.merge(change.getClaim());
+                samhsaUtil.processClaim(change.getClaim(), entityManager);
```
Comment (aschey-forpeople) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:413]: I'm being slightly paranoid here, but we may want to consider doing these inserts in an explicit transaction. Theoretically, if a request comes in right when we're inserting the data, they could get the new claims after they're inserted, but before the SAMHSA tags are created, resulting in a leak. If we do it in a transaction, they shouldn't be able to see the new data until after all the inserts have completed.

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+   *
+   * @param mcsClaim The claim to check.
+   * @return a list of TagDetail objects, one for each SAMHSA code found in the claim.
+   */
+  private Optional<List<TagDetails>> getPossibleMcsSamhsaFields(RdaMcsClaim mcsClaim) {
+    List<TagDetails> entries = new ArrayList<>();
+    LocalDate serviceDate =
+        mcsClaim.getIdrHdrFromDateOfSvc() == null
+            ? LocalDate.parse("1970-01-01")
+            : mcsClaim.getIdrHdrFromDateOfSvc();
```
Comment (aschey-forpeople) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java:194]: Might be worth checking with @alex-dzeda to make sure we're using the right dates. I recall there was some nuance to this when we were discussing this previously.

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+        isSamhsaCode(Optional.ofNullable(fissClaim.getAdmitDiagCode())),
+        "fiss_claims",
+        "admit_diag_code",
+        null,
+        entries,
+        serviceDate,
+        throughDate);
+    for (RdaFissRevenueLine revenueLine : fissClaim.getRevenueLines()) {
+      buildDetails(
+          isSamhsaCode(Optional.ofNullable(revenueLine.getApcHcpcsApc())),
```
Comment (aschey-forpeople) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java:300]: Might be worth a comment here stating that this column should never contain SAMHSA data, but we're checking it in case someone misinterprets this field.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java (hunk 2 above)
Comment (dondevun) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:413]: After a little bit of research, it appears that `executeProcedure` does run in a transaction:

```
Runs a transaction that does not produce a result (usually an insert or update). Ensures that the transaction is properly committed if the procedure completes normally or rolled back if the procedure throws an exception. Internally uses the adapter method ThrowingConsumer.executeAsFunction and discards the null result.
```

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java (hunk 3 above)
Comment (alex-dzeda) [reply apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java:194]: The safest option is to imagine a time period (eg start and end date). If any of those dates is after an applicable start date, I'd consider it to be within range.

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+      String column,
+      Integer lineNum,
+      List<TagDetails> detailsList,
+      LocalDate serviceDate,
+      LocalDate throughDate) {
+    if (entry.isPresent()) {
+      try {
+        LocalDate startDate = LocalDate.parse(entry.get().getStartDate());
+        LocalDate endDate =
+            entry.get().getEndDate().equals("Active")
```
Comment (aschey-forpeople) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java:252]: Maybe make this a case-insensitive comparison just to be safe?

Comment (dondevun) [reply apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java]: Good call

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+      Integer lineNum,
+      List<TagDetails> detailsList,
+      LocalDate serviceDate,
+      LocalDate throughDate) {
+    if (entry.isPresent()) {
+      try {
+        LocalDate startDate = LocalDate.parse(entry.get().getStartDate());
+        LocalDate endDate =
+            entry.get().getEndDate().equalsIgnoreCase("Active")
+                ? LocalDate.now()
```
Comment (aschey-forpeople) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java:253]: I'm not sure if defaulting to `now()` is the safest option for an active code since it would cause issues if the claim through date is in the future for some reason. I'd probably use `LocalDate.MAX` here.

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+                : LocalDate.parse(entry.get().getEndDate());
+
+        // if the throughDate is not between the start and end date,
+        // and the serviceDate is not between the start and end date,
+        // then the claim falls outside the date range of the SAMHSA code.
+        if ((throughDate.isBefore(startDate) || throughDate.isAfter(endDate))
```
Comment (aschey-forpeople) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java:259]: This looks correct, but it's a bit hard to follow. Maybe it would help with readability to make an `isBetween` helper function?

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java (hunk 6 above)
Comment (dondevun) [reply apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java]: Absolutely.

**Review by aschey-forpeople — APPROVED** [review]
Thanks for taking this on, this looks good!
